homeassistant/components/vesync/switch.py

```diff
@@ -45,53 +71,45 @@ def _setup_entities(
     async_add_entities,
     coordinator: VeSyncDataCoordinator,
 ):
-    """Check if device is a switch and add entity."""
-    entities: list[VeSyncBaseSwitch] = []
-    for dev in devices:
-        if DEV_TYPE_TO_HA.get(dev.device_type) == "outlet":
-            entities.append(VeSyncSwitchHA(dev, coordinator))
-        elif DEV_TYPE_TO_HA.get(dev.device_type) == "switch":
-            entities.append(VeSyncLightSwitch(dev, coordinator))
-
-    async_add_entities(entities, update_before_add=True)
+    """Check if device is online and add entity."""
+    async_add_entities(
+        VeSyncSwitchEntity(dev, description, coordinator)
+        for dev in devices
+        for description in SENSOR_DESCRIPTIONS
+        if description.exists_fn(dev)
+    )
 
 
-class VeSyncBaseSwitch(VeSyncBaseEntity, SwitchEntity):
-    """Base class for VeSync switch Device Representations."""
+class VeSyncSwitchEntity(SwitchEntity, VeSyncBaseEntity):
+    """VeSync switch entity class."""
 
-    _attr_name = None
+    entity_description: VeSyncSwitchEntityDescription
 
-    def turn_on(self, **kwargs: Any) -> None:
-        """Turn the device on."""
-        self.device.turn_on()
+    def __init__(
+        self,
+        device: VeSyncBaseDevice,
+        description: VeSyncSwitchEntityDescription,
+        coordinator: VeSyncDataCoordinator,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(device, coordinator)
+        self.entity_description = description
+        if is_outlet(self.device):
+            self._attr_device_class = SwitchDeviceClass.OUTLET
+        if is_wall_switch(self.device):
+            self._attr_device_class = SwitchDeviceClass.SWITCH
 
     @property
-    def is_on(self) -> bool:
-        """Return True if device is on."""
-        return self.device.device_status == "on"
+    def is_on(self) -> bool | None:
+        """Return the entity value to represent the entity state."""
+        return self.entity_description.is_on(self.device)
 
     def turn_off(self, **kwargs: Any) -> None:
-        """Turn the device off."""
+        """Turn the entity off."""
         self.device.turn_off()
+        self.schedule_update_ha_state()
```

Comment:
Shouldn't schedule_update_ha_state be only called if turn_off succeeds?

Reply:
Some other spots have it this way - my theory was after touching devices we should check the source data regardless to ensure we are updated.   Both approaches though in most cases would show correctly I would think.  Which do you prefer?

Reply:
The turn_on/off return true if operation succeeds and I think that should be respected and used as the basis of state refresh.

Reply:
[Switch ](https://github.com/webdjoe/pyvesync/blob/68abac78f9dc311ba6c7d4077ab2ed75ddd54e28/src/pyvesync/vesyncswitch.py#L174 )returns true if the web call succeeded, we should use that return value to schedule_update_ha_state.

Reply:
Adjusted.